include/unifex/task.hpp

```diff
@@ -74,112 +75,109 @@ template <typename T>
+    if (p.info_) {
+      visit_continuations(*p.info_, (Func &&) func);
+    }
+  }
+
+  friend inplace_stop_token tag_invoke(tag_t<get_stop_token>, const _promise_base& p) noexcept {
+    return p.stoken_;
+  }
+
+  using done_callback = coro::coroutine_handle<>(void*) noexcept;
+
+  coro::coroutine_handle<> continuation_;
+  done_callback* doneCallback_ = &default_unhandled_done_callback;
+  std::optional<continuation_info> info_;
+  inplace_stop_token stoken_;
+};
+
 template <typename T>
-struct _promise {
+struct _return_value_or_void {
   struct type {
-    using result_type = T;
+    template(typename Value = T)
+        (requires convertible_to<Value, T> AND constructible_from<T, Value>)
+    void return_value(Value&& value) noexcept(
+        std::is_nothrow_constructible_v<T, Value>) {
+      expected_.reset_value();
+      unifex::activate_union_member(expected_.value_, (Value &&) value);
+      expected_.state_ = _state::value;
+    }
+    _expected<T> expected_;
+  };
+};
 
-    void reset_value() noexcept {
-      switch (std::exchange(state_, state::empty)) {
-        case state::value:
-          unifex::deactivate_union_member(value_);
-          break;
-        case state::exception:
-          unifex::deactivate_union_member(exception_);
-          break;
-        default:
-          break;
-      }
+template <>
+struct _return_value_or_void<void> {
+  struct type {
+    void return_void() noexcept {
+      expected_.reset_value();
+      unifex::activate_union_member(expected_.value_);
+      expected_.state_ = _state::value;
     }
+    _expected<void> expected_;
+  };
+};
+
+template <typename T>
+struct _promise {
+  struct type : _promise_base, _return_value_or_void<T>::type {
+    using result_type = T;
 
     typename _task<T>::type get_return_object() noexcept {
       return typename _task<T>::type{
           coro::coroutine_handle<type>::from_promise(*this)};
     }
 
-    coro::suspend_always initial_suspend() noexcept {
-      return {};
-    }
-
     auto final_suspend() noexcept {
-      struct awaiter {
-        bool await_ready() noexcept {
-          return false;
-        }
+      struct awaiter : _final_suspend_awaiter_base {
         auto await_suspend(coro::coroutine_handle<type> h) noexcept {
           return h.promise().continuation_;
         }
-        void await_resume() noexcept {}
       };
       return awaiter{};
     }
 
-    coro::coroutine_handle<> unhandled_done() noexcept {
-      return doneCallback_(continuation_.address());
-    }
-
     void unhandled_exception() noexcept {
-      reset_value();
-      unifex::activate_union_member(exception_, std::current_exception());
-      state_ = state::exception;
+      this->expected_.reset_value();
+      unifex::activate_union_member(this->expected_.exception_, std::current_exception());
+      this->expected_.state_ = _state::exception;
     }
 
-    template(typename Value)
+    template(typename Value = T)
```

Comment:
The `= T` default shouldn't be here.
The return-type of the coroutine has nothing to do with the type being co_awaited.

Reply:
Ah, stray edit. Thanks. The one I added on `return_value` was the intentional one.